File: habit-service/src/main/java/io/github/xpakx/habittracker/habit/Habit.java
package io.github.xpakx.habittracker.habit;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.*;
import java.time.LocalDateTime;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@NamedEntityGraph(name = "habit-details",
        attributeNodes = {@NamedAttributeNode("context"), @NamedAttributeNode("trigger")}
)
public class Habit {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String name;
    private String description;

    @ManyToOne
    @JoinColumn(name = "context_id")
    @JsonIgnore
    private HabitContext context;

    @OneToOne(cascade = {CascadeType.ALL})
    @JoinColumn(name = "trigger_id")
    @JsonIgnore
    private HabitTrigger trigger;

    private Integer interval;
    private Integer dailyCompletions;
    private Integer completions;
    private LocalDateTime start;
    private LocalDateTime nextDue;
    private Integer difficulty;
}

File: frontend/src/app/habit/dto/habit.ts
import { HabitContextMin } from "./habit-context-min";
import { HabitTrigger } from "./habit-trigger";

export interface Habit {
    id: number;
    name: String;
    description: String;
    context: HabitContextMin;
    trigger: HabitTrigger;

    interval: number;
    dailyCompletions: number;
    start: Date;
    nextDue: Date;
}